### PR TITLE
[WIP] Provide UICollectionView category to apply diff results

### PR DESCRIPTION
**Pull request by rnystrom**
## Changes in this pull request

@jessesquires is this the sort of thing you had in mind for #870?

I started taking some notes on how to merge this all together so that there's a single entry point into batch updating a `UICollectionView` so that this can be shared w/ `IGListAdapterUpdater`. That's my preferred approach, but it's a bit complicated. If we're going to get more serious about this, I'll explore that path. 

My biggest fear is that we wont use this category, and that supporting it will be a monster PITA. If I can centralize the codepaths then it'll be supported "for free".

### Checklist

- [ ] All tests pass. Demo project builds and runs.
- [ ] I added tests, an experiment, or detailed why my change isn't tested.
- [ ] I added an entry to the `CHANGELOG.md` for any breaking changes, enhancements, or bug fixes.

### CODE REVIEW

**Comment by rnystrom** [comment]
Not much activity, gonna let this die out. Can revive if this becomes more requested.

**Comment by DreamingInBinary** [comment]
Hey hey @rnystrom - totally understand closing this one down. For our use case at Buffer this one is key, so I'm trying to roll it on my own. I've got a few questions, if that's cool.

1) By diving in more and more, it seems any updates should manifest themselves as deletes/inserts in `performBatchUpdates` - is that right? At first, I was allocating a `IGListBatchUpdateData` object for batch updates but hit exceptions are reloading items at index paths.

2) Our use case is diffing multiple sections, never just one. My flow right now is, diff the sections, loop through any updates that comes from it and snag a `IGListIndexPathResult` for each of them. Is that the sane way of thinking about it?

3) Just a quick one, in Apple's sample code I've come across a comment that mentions the "correct" way to update things in batch updates was Delete/Insert/Reload/Move. Though I can't specifically remember if this example was a table view, which is possible. In this code I noticed its Delete/Insert/Move.

I've got the exact code we're rolling with right now (some Buffer specific objects) - wondered if I'm on a good path forward? Totally no rush or expectation to respond, just thought I would pick your brain if possible 😄 

<hr /> 

```
- (void)reloadByDiffingOldUpdates:(NSArray<UpdatesCollectionSectionModel *> *)oldUpdates newUpdates:(NSArray<UpdatesCollectionSectionModel *> *)newUpdates diffCompleted:(void (^)(BOOL))diffCompleted completion:(void (^)(BOOL))completion {
    
    // Represents if there were any changes at all within sections
    IGListIndexSetResult *diffs = IGListDiff(oldUpdates, newUpdates, IGListDiffEquality);

    // Early out
    if (!diffs || !diffs.hasChanges) {
        if (diffCompleted) {
            diffCompleted(YES);
        }
        if(completion){
            completion(YES);
        }
        return;
    }
    
    // Now get the granular changes within each section
    NSMutableIndexSet *inserts = [diffs.inserts mutableCopy];
    NSMutableIndexSet *deletes = [diffs.deletes mutableCopy];
    NSSet *moves = [[NSSet alloc] initWithArray:diffs.moves];
    NSMutableIndexSet *updates = [diffs.updates mutableCopy];
    
    // Changes within the section, if applicable
    NSMutableArray<NSIndexPath *> *itemInserts = [NSMutableArray new];
    NSMutableArray<NSIndexPath *> *itemDeletes = [NSMutableArray new];
    NSMutableArray<IGListMoveIndexPath *> *itemMoves = [NSMutableArray new];
    
    // A section was updated, let's find out what happened
    [updates enumerateIndexesUsingBlock:^(NSUInteger oldIndex, BOOL *stop) {
        __block NSUInteger newIndex = NSNotFound;
        
        // Assign any old sections that were entirely moved and have a new section index? I might be off here
        [moves enumerateObjectsUsingBlock:^(IGListMoveIndex *moveIndex, BOOL *moveStop) {
            if (moveIndex.from == oldIndex) {
                newIndex = moveIndex.to;
                *moveStop = YES;
            }
        }];

        if (newIndex == NSNotFound) newIndex = oldIndex;
        
        IGListIndexPathResult *paths = IGListDiffPaths(oldIndex, newIndex, oldUpdates[oldIndex].updates, newUpdates[newIndex].updates, IGListDiffEquality);
        paths = [paths resultForBatchUpdates];
        [itemInserts addObjectsFromArray:paths.inserts];
        [itemDeletes addObjectsFromArray:paths.deletes];
        [itemMoves addObjectsFromArray:paths.moves];
    }];
    
    diffCompleted(YES);

    [self performBatchUpdates:^{
        
        // Sections
        [self insertSections:inserts];
        [self deleteSections:deletes];
        for (IGListMoveIndex *move in moves) {
            [self moveSection:move.from toSection:move.to];
        }
        
        // Index paths
        [self insertItemsAtIndexPaths:itemInserts];
        [self deleteItemsAtIndexPaths:itemDeletes];
        for (IGListMoveIndexPath *move in itemMoves) {
            [self moveItemAtIndexPath:move.from toIndexPath:move.to];
        }
        
    } completion:^ (BOOL done) {
        completion(YES);
    }];
}
```

**Comment by rnystrom** [comment]
This is awesome! I'm glad someone (esp Buffer ❤️ ) is interested in this. I'll try and be as helpful as I can! Also very happy to make this a part of IGLK.

### 1.

> By diving in more and more, it seems any updates should manifest themselves as deletes/inserts in performBatchUpdates - is that right? At first, I was allocating a IGListBatchUpdateData object for batch updates but hit exceptions around reloading items at index paths (i.e. same index path was both a move and reload).

Yup, deletes+inserts+moves. 

I spent a **lot** of time investigating reloads. Originally reloads were supported, however at Instagram's scale we detected a significant number of crashes with exceptions like:

```
*** Terminating app due to uncaught exception 'NSInternalInconsistencyException', reason: 'attempt to perform an insert and a move to the same section (4)'
```

`UICollectionView`'s documentation is misleading:

> You can use this method in cases where you want to make multiple changes to the collection view in one single animated operation, as opposed to in several separate animations. You might use this method to **insert, delete, reload, or move** cells or use it to change the layout parameters associated with one or more cells.

This is true, but reload transitions are implemented strangely.

To prove the inconsistency in messaging, the delete+insert+move API [docs](https://developer.apple.com/documentation/uikit/uicollectionview/1618097-insertitemsatindexpaths?language=objc) have the line:

> You can also call this method from a block passed to the performBatchUpdates:completion: method ...

This line is curiously missing from [reload's docs](https://developer.apple.com/documentation/uikit/uicollectionview/1618055-reloaditemsatindexpaths?language=objc).

We've opened a radar which you can view (and dupe! 🙏 ) with #690. It includes a very basic sample app that repros this error. With this finding we decided that reloads inside batch updates are considered unstable and shouldn't be used.

#### 2.

> Our use case is diffing multiple sections, never just one. My flow right now is, diff the sections, loop through any updates that comes from it and snag a IGListIndexPathResult for each of them via its resultForBatchUpdates. Is that the sane way of thinking about it?

Yup, and your code is handling when sections are repositioned too, so you should be ok. Keep in mind that this is an _n_<sup>2</sup> operation tho. I've always wanted to write an algo that could take in 2d arrays, flatten to 1d, diff, then output a 2d result. Saving that for a rainy day (or someone else to contribute 😇 ).

But don't get bogged down w/ perf. If its a problem, diff on a bg queue. But as long as you aren't dealing w/ thousands of sections+items, you're fine.

Also, `UICollectionView` changes are orders-of-magnitude more expensive than our diffing algo. We're still testing the improvements from falling back to `reloadData` for monster changesets in 61c1524bb0c59d5eb43f65b02442ad50dfb2d0dc.

#### 3.

Basically see 1.

---

Some notes on the code:

- You probably want to use `resultForBatchUpdates` on the section changes too, those will have reloads as-is
- Some section+item ops have issues. See #691 for another radar. See some of the cleaning in [IGListBatchUpdateData](https://github.com/Instagram/IGListKit/blob/master/Source/Common/IGListBatchUpdateData.mm).
  - I was actually going to update this class to take in reloads as well and do the conversion internally instead of where it is now in `IGListAdapterUpdater`. That way you could use the update data standalone. I'd accept a PR for this!

**Comment by DreamingInBinary** [comment]
Wow @rnystrom this info is gold 💯 - huge thanks as we're now able to land this change in our iPhone X release! Woop woop 😄 

All of that is great insight, and performance has been stellar so for us too as we rarely hit a collection view approaching 1k items within it. Thanks for all of your work with good ol' collection view, I know it can be a temperamental one sometimes haha.

Also, I've got both of those radars duped as well. Looking forward to more IGListKit goodness in the future 😄
